Include information about borrowed values in command.md

IntegralPilot opened the pull request:
This pull request updates the part of command.md about asynchronous functions to include an important note about how borrowed values like &str are not allowed at present, and updates the code sample to include an example of using a String instead as a workaround.

Helps resolve `https://github.com/tauri-apps/tauri/issues/6733`.

IntegralPilot committed:
Include information about borrowed values in command.md

Helps resolve https://github.com/tauri-apps/tauri/issues/6733

FabianLars commented:
fwiw here's a similar PR: <https://github.com/tauri-apps/tauri-docs/pull/1163> - i'll take a look as soon as i can.

IntegralPilot commented:
oops... made a typo in the word Asynchronous, trying to figure out how to add a new commit in the same PR

*edit: figured out, all is fixed now*

IntegralPilot committed:
Update command.md with better grammar

IntegralPilot committed:
Fix another grammatical mistake in this PR

IntegralPilot commented:
Fixed grammatical mistakes! Also changed the page title for command.md from _Calling Rust from the frontend_ to _Communication between Rust and the Frontend_ as it appears more relevant.

IntegralPilot commented:
Okay! Just read another issue on Tauri main, and it turns out this effects more than just types with the & like my documentation suggests. I'm going to fix that now, and maybe include an example of the result wrapping...

*done! also prettified some of my markdown formatting, i'll just check after the build, but I think this is nearly ready to merge!*

IntegralPilot committed:
Expand on async borrowing fix in Command.md

IntegralPilot commented:
Build checked, I'm really happy with my changes, I think this PR is finalised from my end now. 

Hopefully, when reviewed and merged, this will help save people some debugging time on the borrowing-async issue.

FabianLars commented on docs/guides/features/command.md:
```diff
@@ -1,4 +1,4 @@
-# Calling Rust from the frontend
+# Communication between Rust and the Frontend
```
```suggestion
# Calling Rust from the frontend
```

Please revert this for now as it doesn't look good in the sidebar since it's the only item taking up multiple lines:
![grafik](https://user-images.githubusercontent.com/30730186/232746267-359644e8-f5bf-4882-8e30-b31a71f398a8.png)

We could revisit this in the new docs in the future.

IntegralPilot replied:
Yep, didn't notice that, thanks. Changing title back now...

IntegralPilot replied:
Committed!

FabianLars commented on docs/guides/features/command.md:
```diff
@@ -164,28 +164,61 @@ It also gives you full control over the way your error type gets serialized. In
+
+:::note
+
+You need to be careful when creating asynchronous functions using Tauri. Currently, you cannot include types that are borrowed, such as those with the & symbol, in the constructor of an asynchronous function. Some common examples of types like this are `&str` and `State<'_, Data>`.
```
```suggestion
You need to be careful when creating asynchronous functions using Tauri. Currently, you cannot include borrowed arguments in the signature of an asynchronous function. Some common examples of types like this are `&str` and `State<'_, Data>`.
```

My main issue with this one is `such as those with the & symbol` and `constructor`. "constructor" is just wrong in this context afaik, and "such as those with the & symbol" sounds a bit too weird imo, even though i see how that would help beginners understand what we're talking about.

IntegralPilot replied:
Definitely, sounds a lot better than fluid. I think `constructor` is the name for function signatures in Kotlin, got my language lingo a bit mixed up there, thanks.

IntegralPilot replied:
Commited!

FabianLars commented on docs/guides/features/command.md:
```diff
@@ -164,28 +164,61 @@ It also gives you full control over the way your error type gets serialized. In
+:::
+
+
+If your command needs to run asynchronously, simply declare it as `async`, and resolve any issues with borrowed types.
+
+When working with borrowed types, there are two main options.
```
```suggestion
If your command needs to run asynchronously, simply declare it as `async`.

When working with borrowed types, you have to make additional changes. These are your two main options:
```

Not quite happy with my suggestion but i hope you're getting the idea.

IntegralPilot replied:
I think the suggestion is great, makes it more fluid. :D

IntegralPilot replied:
Committed!

FabianLars commented on docs/guides/features/command.md:
```diff
@@ -164,28 +164,61 @@ It also gives you full control over the way your error type gets serialized. In
+
+When working with borrowed types, there are two main options.
+
+**Option 1**: Convert the type, such as `&str` to a similar type that is not borrowed, such as `String`. This won't work for all types.
```
```suggestion
**Option 1**: Convert the type, such as `&str` to a similar type that is not borrowed, such as `String`. This may not work for all types.
```

IntegralPilot replied:
I think may not works better, thanks. Committed!

FabianLars commented on docs/guides/features/command.md:
````diff
@@ -164,28 +164,61 @@ It also gives you full control over the way your error type gets serialized. In
+*Example:*
+
+```rust
+// Declare the constructor using String instead of &str, as &str is borrowed and thus unsupported
````
```suggestion
// Declare the async function using String instead of &str, as &str is borrowed and thus unsupported
```

IntegralPilot replied:
Again, got my Kotlin and my Rust lingo mixed up, constructor definitely doesn't fit well there, thanks. Committed!

FabianLars commented on docs/guides/features/command.md:
````diff
@@ -164,28 +164,61 @@ It also gives you full control over the way your error type gets serialized. In
+}
+```
+
+**Option 2**: Wrap the return type in a Result. This one is a bit harder to implement, but works for all types.
````
```suggestion
**Option 2**: Wrap the return type in a Result. This one is a bit harder to implement, but should work for all types.
```

IntegralPilot replied:
Again, should work sounds a lot better. Committed! Thanks

FabianLars commented on docs/guides/features/command.md:
```diff
@@ -164,28 +164,61 @@ It also gives you full control over the way your error type gets serialized. In
+
+**Option 2**: Wrap the return type in a Result. This one is a bit harder to implement, but works for all types.
+
+If you have no return type, like in our function, use the return type `Result<(), ()>`. Otherwise, use the return type `Result<bool, ()>`, replacing `bool` with whatever type you wish to return.
```
I think we need to be a bit more explicit that this is only if you don't want to return any actual errors from the command.

IntegralPilot replied:
Hmm... I've fixed that just now, but I think my English is a bit clunky.

FabianLars commented on docs/guides/features/command.md:
```diff
@@ -164,28 +164,61 @@ It also gives you full control over the way your error type gets serialized. In
 
-If your command needs to run asynchronously, simply declare it as `async`:
+
+:::note
```
```suggestion
:::warning
```

Let it take the user's attention :D

IntegralPilot replied:
:D

FabianLars commented on docs/guides/features/command.md:
````diff
@@ -164,28 +164,61 @@ It also gives you full control over the way your error type gets serialized. In
 
 ```js
-invoke('my_custom_command').then(() => console.log('Completed!'))
+invoke('my_custom_command', {value: "Hello, Async!"}).then(() => console.log('Completed!'))
````
```suggestion
invoke('my_custom_command', { value: "Hello, Async!" }).then(() => console.log('Completed!'))
```
I know that's really picky but i have to 😂

IntegralPilot replied:
It does look nicer...

FabianLars commented on docs/guides/features/command.md:
````diff
@@ -164,28 +164,61 @@ It also gives you full control over the way your error type gets serialized. In
 ```
 
-Since invoking the command from JS already returns a promise, it works just like any other command:
+**Invoking from JS:** Since invoking the command from JS already returns a promise, it works just like any other command:
````
```suggestion
#### Invoking from JS

Since invoking the command from JS already returns a promise, it works just like any other command:
```

IntegralPilot replied:
That markdown looks better, thanks.

FabianLars reviewed (changes requested):
Thanks for your contribution!

While this is imo is almost good to merge as-is i still went ahead and added a few pretty picky comments on stuff i'd like to talk about. Please take a look at them and see what makes sense to you and what doesn't (not all of my suggestions must be applied!)

IntegralPilot committed:
Revert title in command.md

Co-authored-by: Fabian-Lars <fabianlars@fabianlars.de>

IntegralPilot committed:
Update docs/guides/features/command.md

Co-authored-by: Fabian-Lars <fabianlars@fabianlars.de>

IntegralPilot committed:
Update docs/guides/features/command.md

Co-authored-by: Fabian-Lars <fabianlars@fabianlars.de>

IntegralPilot committed:
Update docs/guides/features/command.md

Co-authored-by: Fabian-Lars <fabianlars@fabianlars.de>

IntegralPilot committed:
Update docs/guides/features/command.md

Co-authored-by: Fabian-Lars <fabianlars@fabianlars.de>

IntegralPilot committed:
Update docs/guides/features/command.md

Co-authored-by: Fabian-Lars <fabianlars@fabianlars.de>

IntegralPilot committed:
Update docs/guides/features/command.md

Co-authored-by: Fabian-Lars <fabianlars@fabianlars.de>

IntegralPilot committed:
Update docs/guides/features/command.md

Co-authored-by: Fabian-Lars <fabianlars@fabianlars.de>

IntegralPilot committed:
Update docs/guides/features/command.md

Co-authored-by: Fabian-Lars <fabianlars@fabianlars.de>

IntegralPilot committed:
Explain return types better

IntegralPilot commented:
Thanks for all your help! I've applied all your changes, I think they were great. I've tried to explain the return types a bit better, but I think my English was a bit clunky.

FabianLars committed:
some typos and stuff

FabianLars reviewed (approved):
Thanks again!!

FabianLars committed:
move some lines around

IntegralPilot commented:
Wow, just read the latest commits from you before merge, and they are great!

Thank you so much for helping to make this PR so much better, its being lots of fun contributing back to one of the open source projects I love.
